### PR TITLE
Recompact blocks with large number of tombstones

### PR DESCRIPTION
**Pull request by Gouthamve**
Carried over #91 as it diverged too much from master.

### CODE REVIEW

File: compact.go [hunk 1]
```diff
@@ -439,6 +455,9 @@ func populateBlock(blocks []Block, indexw IndexWriter, chunkw ChunkWriter) (*Blo
 
 		meta.Stats.NumChunks += uint64(len(chks))
 		meta.Stats.NumSeries++
+		for _, chk := range chks {
+			meta.Stats.NumSamples += uint64(binary.BigEndian.Uint16(chk.Chunk.Bytes()))
```
Comment (fabxc) [inline compact.go]: We should eventually make this a first class property of the chunk interface. Strictly speaking not all compression formats would need this. But enforcing it also has negligible cost. In doubt one could have the `Size()` function just iterate through all elements.

Comment (fabxc) [inline compact.go]: Actually, maybe we should just do it as part of this PR.

File: compact.go [hunk 2]
```diff
@@ -160,7 +161,19 @@ func (c *compactor) Plan() ([][]string, error) {
...
+	if len(dirs) > 1 {
+		return dirs, nil
+	}
+
+	// Compact any blocks that have >5% tombstones.
+	for i := len(dms) - 1; i >= 0; i-- {
+		if dms[i].meta.Stats.NumSeries/dms[i].meta.Stats.NumTombstones <= 20 { // 5%
+			return [][]string{dms[i].dir}, nil
+		}
+	}
```
Comment (fabxc) [inline compact.go]: Didn't you have something in there that only made us consider blocks with a relatively long time range? 
I still that still makes sense to have as short ones will be compacted soon enough.

Comment (Gouthamve) [inline compact.go]: Added a rough heuristic now.

File: chunks/chunk.go [hunk 3]
```diff
@@ -43,6 +43,7 @@ type Chunk interface {
 	Encoding() Encoding
 	Appender() (Appender, error)
 	Iterator() Iterator
+	NumSamples() int64
```
Comment (fabxc) [inline chunks/chunk.go]: Just `int` seems fine here.

Comment (Gouthamve) [inline chunks/chunk.go]: Done, PTAL.

**Comment by fabxc** [comment]
👍

**Pull request merged**